text/0182-otlp-remote-parent.md

```diff
@@ -0,0 +1,108 @@
+# Export SpanContext.IsRemote in OTLP
+
+Update OTLP to indicate whether a span's parent is remote.
+
+## Motivation
+
+It is sometimes useful to post-process or visualise only entry-point spans: spans which either have no parent (trace roots), or which have a remote parent.
+For example, the Elastic APM solution highlights entry-point spans (Elastic APM refers to these as "transactions") and surfaces these as top-level operations
+in its user interface.
+
+The goal is to identify the spans which represent a request that is entering a service, or originating within a service, without having to first assemble the
+complete distributed trace as a DAG (Directed Acyclic Graph). It is trivially possible to identify trace roots, but it is not possible to identify spans with
+remote parents.
```

Comment:
I think in the context of this OTEP, it's worth clarifying why the existing `SpanKind` cannot be use to achieve the desired outcome.

This is text from the [`SpanKind` specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind):
> The first property described by `SpanKind` reflects whether the Span is a "logical" remote child or parent. By "logical", we mean that the span is logically a remote child or parent, from the point of view of the library that is being instrumented. Spans with a remote parent are interesting because they are sources of external load. Spans with a remote child are interesting because they reflect a non-local system dependency.

Given this and the example below, couldn't one treat every `SERVER` and `CONSUMER` span as spans with a remote parent? If not, then I think this OTEP should also propose an adapted specification of  `SpanKind`. Otherwise the specification would suggest that a span has both `SpanKind` which reflects whether the span is a "logical" remote child or parent, and `ParentSpanIsRemote`, which also reflects whether the span is a remote child.

Reply:
Unfortunately, the specification on `SpanKind` remains unclear for the `CONSUMER` SpanKind, as you can see [in the summary table](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind):

> <img width="786" alt="image" src="https://user-images.githubusercontent.com/866830/216545008-ce591497-9395-4167-b7b2-e7cca82718a7.png">

`Remote Incoming` is a `maybe` for a `CONSUMER` span. So, to solve that ambiguity I see the following alternatives:

- Introduce a new attribute `parent_span_is_remote` and API `SpanContext.HasRemoteParent` as proposed in this OTEP. (And I agree that would potentially require to adapt / refine the `SpanKind` specification as well to clarify the relation between `SpanKind` and `parent_span_is_remote`.)
- Revise the values for `SpanKind` (especially for `CONSUMER`) to make it unambiguous wrt. `Remote Incoming`. However, this would be a breaking change for spans that are reported as of now as `CONSUMER` spans.

Reply:
Would you be fine adding the second alternative as an alternative solution to the OTEP? And also adding the need to clarify the relationship between `SpanKind` and `parent_span_is_remote` as an open question?

Reply:
@pyohannes Thanks, added as requested.

Reply:
I think "remote incoming" is always "maybe" for Server and Consumer. E.g. you can have nested server spans in the same process for different framework levels

(Aside: Sync/Async also cannot be determined by the parent, it depends on what the client/producer does)

Reply:
@Oberon00 Are you saying that `SpanKind` is not suitable to encode the `hasRemoteParent` information? And that we should rather make it explicit with a new `hasRemoteParent` in the API?

Reply:
> Are you saying that SpanKind is not suitable to encode the hasRemoteParent information?

Yes, SpanKing is definitely not suitable for this (it could be used as input for a heuristic at best). That is is also what the original comment in this thread is saying